core/sched.c

```diff
@@ -281,6 +281,14 @@ void sched_set_status(thread_t *process, thread_status_t status)
 void sched_switch(uint16_t other_prio)
 {
     thread_t *active_thread = thread_get_active();
+
+    /* If a thread exists and no other thread is runnable, we may end up in
+     * a situation with no active thread. This can not occur if there is an
+     * idle thread, which is always runnable, though */
+    if (!IS_USED(MODULE_CORE_IDLE_THREAD) && unlikely(active_thread == NULL)) {
```

Comment:
This change is correct, bit I'm not sure this is holding it right. `thread_yield_higher()` can always be used and is actually the preferred choice. `sched_switch()` was an optimization path for cases like unlocking a mutex or waking up another thread, *from some thread*, to only invoke the full scheduler if needed (as the target thread / priority is basically known).

If this is actually called from ISRs, that is the only way there is no `active_thread`. It might make more sense to encode this properly. As in, if there is no active thread, the `irq_is_in()` below becomes redundant, that is implicit. And in that case, `sched_context_switch_request = 1` might be the correct thing to do.
Or, if `irq_is_in()`, always trigger scheduler, otherwise, assume not in ISR and that there is an `active_thread`.

(In theory, this function can be used to speed up task switching by quite a bit for some cases, b/c we could directly switch to the receiving thread, skipping register storing for the calling thread. we never implemented that, though.)

Reply:
> `sched_switch()` was an optimization path for cases like unlocking a mutex or waking up another thread, from some thread, to only invoke the full scheduler if needed (as the target thread / priority is basically known).

This makes it sound like an ugly hack / premature optimization. In normal workloads contention over a mutex is pretty unlikely and the hot path is the case a `mutex_unlock()` will not find any waiters. And for real time behavior: Adding a `sched_switch()` prior to the call to `thread_yield_higher()` increases the worst case latency at the hope to occasionally omit the call to `thread_yield_higher()`.

I wonder if just getting rid of `sched_switch()` is the better approach here. See https://github.com/RIOT-OS/RIOT/pull/20890 for removing it for `mutex_unlock()`. There are a few other calls to `sched_switch()` that I think might happen from IRQ context to also address to actually fix the issue.